File: dotnet/lesson-1/Store.cs
// Build, cache, and load the document index.
// Mirrors localrag/store.py: a JSON file of chunks (plus a (path, mtime, size)
// fingerprint) under the cache dir. BM25 only — no vectors in this C# port.

using System.Text.Json;
using System.Text.Json.Serialization;

namespace LocalRag;

public record FileFingerprint(
    [property: JsonPropertyName("path")] string Path,
    [property: JsonPropertyName("mtime")] long Mtime,
    [property: JsonPropertyName("size")] long Size);

public record IndexFile(
    [property: JsonPropertyName("fingerprint")] List<FileFingerprint> Fingerprint,
    [property: JsonPropertyName("chunks")] List<ChunkRecord> Chunks);

// JSON shape matches Python's chunk dicts (snake_case keys).
public record ChunkRecord(
    [property: JsonPropertyName("source")] string Source,
    [property: JsonPropertyName("page_number")] int PageNumber,
    [property: JsonPropertyName("chunk_index")] int ChunkIndex,
    [property: JsonPropertyName("text")] string Text);

public static class Store
{
    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
    };

    private static string IndexPath(Config config) => Path.Combine(config.CacheDir, "index.json");

    private static List<FileFingerprint> Fingerprint(List<string> files)
    {
        var fp = new List<FileFingerprint>();
        foreach (var p in files)
        {
            var info = new FileInfo(p);
            var mtime = ((DateTimeOffset)info.LastWriteTimeUtc).ToUnixTimeSeconds();
            fp.Add(new FileFingerprint(p, mtime, info.Length));
        }
        return fp;
    }

    private static bool FingerprintsEqual(List<FileFingerprint> a, List<FileFingerprint> b)
    {
        if (a.Count != b.Count)
        {
            return false;
        }
        for (var i = 0; i < a.Count; i++)
        {
            if (a[i].Path != b[i].Path || a[i].Mtime != b[i].Mtime || a[i].Size != b[i].Size)
            {
                return false;
            }
        }
        return true;
    }

    /// <summary>True if the cache is missing or the docs folder changed since last build.</summary>
    public static bool IsStale(Config config)
    {
        var path = IndexPath(config);
        if (!File.Exists(path))
        {
            return true;
        }
        IndexFile? data;
        try
        {
            data = JsonSerializer.Deserialize<IndexFile>(File.ReadAllText(path), JsonOpts);
        }
        catch
        {
            return true;
        }
        if (data is null)
        {
            return true;
        }
        return !FingerprintsEqual(data.Fingerprint, Fingerprint(Extract.DiscoverFiles(config.DocsDir)));
    }

    /// <summary>Extract + chunk every supported file. Returns (chunks, file_count).</summary>
    public static (List<Chunk> Chunks, int FileCount) BuildIndex(Config config)
    {
        var files = Extract.DiscoverFiles(config.DocsDir);
        var chunks = new List<Chunk>();
        foreach (var path in files)
        {
            chunks.AddRange(Chunking.ChunkPages(Extract.ExtractPages(path)));
        }

        Directory.CreateDirectory(config.CacheDir);
        var records = chunks
            .Select(c => new ChunkRecord(c.Source, c.PageNumber, c.ChunkIndex, c.Text))
            .ToList();
        var index = new IndexFile(Fingerprint(files), records);
        File.WriteAllText(IndexPath(config), JsonSerializer.Serialize(index, JsonOpts));
        return (chunks, files.Count);
    }

    public static List<Chunk> LoadChunks(Config config)
    {
        var data = JsonSerializer.Deserialize<IndexFile>(File.ReadAllText(IndexPath(config)), JsonOpts)
                   ?? throw new InvalidOperationException("Could not read the index file.");
        return data.Chunks
            .Select(c => new Chunk(c.Source, c.PageNumber, c.ChunkIndex, c.Text))
            .ToList();
    }
}

File: node/lesson-1/src/store.js
// Build, cache, and load the document index.
//
// The index is just a JSON file of chunks (plus metadata) under the cache dir.
// Files are fingerprinted by (path, mtime, size) so re-indexing only happens
// when something changed. Mirrors localrag/store.py (BM25 only — no vectors).

import fs from "node:fs";
import path from "node:path";

import { chunkPages } from "./chunk.js";
import { discoverFiles, extractPages } from "./extract.js";

function fingerprint(files) {
  return files.map((p) => {
    const st = fs.statSync(p);
    return { path: p, mtime: Math.floor(st.mtimeMs / 1000), size: st.size };
  });
}

function indexPath(config) {
  return path.join(config.cacheDir, "index.json");
}

function fingerprintsEqual(a, b) {
  return JSON.stringify(a) === JSON.stringify(b);
}

export function isStale(config) {
  // True if the cache is missing or the docs folder changed since last build.
  const ip = indexPath(config);
  if (!fs.existsSync(ip)) return true;
  let data;
  try {
    data = JSON.parse(fs.readFileSync(ip, "utf-8"));
  } catch {
    return true;
  }
  return !fingerprintsEqual(data.fingerprint, fingerprint(discoverFiles(config.docsDir)));
}

export async function buildIndex(config) {
  // Extract + chunk every supported file. Returns { chunks, fileCount }.
  const files = discoverFiles(config.docsDir);
  const chunks = [];
  for (const filePath of files) {
    const pages = await extractPages(filePath);
    chunks.push(...chunkPages(pages));
  }

  fs.mkdirSync(config.cacheDir, { recursive: true });
  fs.writeFileSync(
    indexPath(config),
    JSON.stringify({ fingerprint: fingerprint(files), chunks }),
    "utf-8"
  );
  return { chunks, fileCount: files.length };
}

export function loadChunks(config) {
  const data = JSON.parse(fs.readFileSync(indexPath(config), "utf-8"));
  return data.chunks;
}
